src/java/org/apache/cassandra/db/VirtualTable.java

```diff
@@ -0,0 +1,148 @@
+import org.apache.cassandra.db.rows.Row;
+import org.apache.cassandra.exceptions.CassandraException;
+import org.apache.cassandra.exceptions.ConfigurationException;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.transport.messages.ResultMessage;
+import org.apache.cassandra.utils.FBUtilities;
+
+import com.google.common.collect.Lists;
+
+/**
+ * Base requirements for a VirtualTable. This is required to provide metadata about the virtual table, such as the
+ * partition and clustering keys, and provide a ReadQuery for a SelectStatement.
+ */
+public abstract class VirtualTable
+{
+    protected final TableMetadata metadata;
+    protected String keyspace;
+    protected String name;
+
+    public VirtualTable(TableMetadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    public String getTableName()
+    {
+        return this.metadata.name;
+    }
+
+    /**
+     * Is this table writable?
+     *
+     * @return True if UPDATE is supported
+     */
+    public boolean writable()
+    {
+        return false;
+    }
+
+    /**
+     * If the table allows unrestricted queries (ie filter on clustering key with no partition). Since These tables are
+     * not backed by the C* data model, this restriction isnt always necessary.
+     */
+    public boolean allowFiltering()
+    {
+        return true;
+    }
+
+    /**
+     * Return some implementation of a ReadQuery for a given select statement and query options.
+     * 
+     * @param selectStatement
+     * @param options
+     * @param limits
+     * @param nowInSec
+     * @return ReadQuery
+     */
+    public abstract ReadQuery getQuery(SelectStatement selectStatement, QueryOptions options, DataLimits limits,
+            int nowInSec);
+
+    /**
+     * Execute an update operation.
+     *
+     * @param partitionKey
+     *            partition key for the update.
+     * @param params
+     *            parameters of the update.
+     */
+    public void mutate(DecoratedKey partitionKey, Row row) throws CassandraException
+    {
+        // this should not be called unless writable is overridden
+        throw new InvalidRequestException("Not Implemented");
+    }
+
+    public static Class<? extends VirtualTable> classFromName(String name)
+    {
+        String className = name.contains(".")
+                ? name
+                : "org.apache.cassandra.db.virtual." + name;
+        Class<VirtualTable> strategyClass = FBUtilities.classForName(className, "virtual table");
+
+        if (!VirtualTable.class.isAssignableFrom(strategyClass))
+        {
+            throw new ConfigurationException(format("Compaction strategy class %s is not derived from VirtualTable",
+                    className));
+        }
+
+        return strategyClass;
+    }
+
+    public static Key createKey()
+    {
+        return new Key();
```

Comment:
Does it make sense to allow empty key creation as API? If not, switch it to a builder so's it can be validated.